File: 2019-2020_programmers/_01_/src/ts/an_unfulfilled_player.ts
const participants = [
  ["leo", "kiki", "eden"],
  ['marina', 'josipa', 'nikola', 'vinko', 'filipa'],
  ['mislav', 'stanko', 'mislav', 'ana']
]

const completions = [
  ['eden', 'kiki'],
  ['josipa', 'filipa', 'marina', 'nikola'],
  ['stanko', 'ana', 'mislav']
]

const solution
  = (_participants: string[], _completions: string[]) => {

    const completionMap
      = _completions.reduce((acc: Map<string, number>, curr: string) => {
        if (!acc.has(curr)) {
          acc.set(curr, 1)
        } else {
          acc.set(curr, acc.get(curr)! + 1)
        }
        return acc
      }, new Map<string, number>())
    
      for (let participant of _participants) {
        if (!completionMap.has(participant)) {
            return participant
        }
      }
  }

console.log(solution(['leo', 'eden', 'kiki'], ['eden', 'kiki']))
console.log(solution(participants[1], completions[1]))
console.log(solution(participants[2], completions[2]))

File: 2019-2020_programmers/_01_/src/rust/an_unfulfilled_player.rs
use std::collections::HashMap;

#[allow(dead_code)]
fn main() {
    println!("{}", solution2(&[ "mislav", "stanko", "mislav", "ana" ], &[ "stanko", "ana", "mislav" ]));
    println!("{}", solution2(&[ "marina", "josipa", "nikola", "vinko", "filipa"	], &["josipa", "filipa", "marina", "nikola"]));
    println!("{}", solution2(&[ "mislav", "stanko", "mislav", "ana"]	, &["stanko", "ana", "mislav"]));
}

pub fn solution2(_participant: &[&str], _completion: &[&str]) -> String {
    let mut p_map: HashMap<&str, i32>= HashMap::new();
    for p in _participant.iter() {
        if p_map.contains_key(p) {
            *p_map.get_mut(p).unwrap() += 1;
        } else {
            p_map.insert(p, 1);
        }
    }
    println!("# participants \n\t{:?}\n", p_map);

    for c in _completion.iter() {
        *p_map.get_mut(c).unwrap() -= 1;
    }
    println!("# completors \n\t{:?}\n", p_map);

    p_map.retain(|&_k, &mut v| v > 0);

    println!("# unfulfilled\n\t{:?}\n", p_map);

    return p_map.iter().fold(String::new(), |acc, (&k,_) | { acc + &k.to_string() });
}
